### PR TITLE
Feature/libcint wrapper

### PR DESCRIPTION
**Pull request by lelemmen**
This PR will address #10, implementing the interface/wrapper to the [Libcint](https://github.com/sunqm/libcint) integral engine.

### CODE REVIEW

File: include/Basis/LibcintInterfacer.hpp [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+
+
+}  // namespace libcint
+
+
+
+
+
+using Libint1eFunction = std::function<int (double*, int*, int*, int, int*, int, double*)>;
```
Comment (tmhuysen) [inline include/Basis/LibcintInterfacer.hpp]:
```suggestion
using Libcint1eFunction = std::function<int (double*, int*, int*, int, int*, int, double*)>;
```

File: include/Basis/LibcintInterfacer.hpp [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+
+/*
+ *  WRAPPERS AROUND C-STYLE LIBCINT ARRAYS
+ */
+namespace libcint {
+
+
+static constexpr int atm_slots = ATM_SLOTS;
+static constexpr int bas_slots = BAS_SLOTS;
```
Comment (tmhuysen) [inline include/Basis/LibcintInterfacer.hpp]: Idem, + Also why is `bas_slots ` declared while `BAS_SLOTS ` is still used in the gqcp code.

File: src/Basis/LibcintInterfacer.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        }
+
+
+        // If there's a new atom, increment the index
+        auto current_atom = current_shell.get_atom();
+        if (current_atom != previous_atom) {
+            atom_index++;
+            previous_atom = current_atom;
+        }
+        raw_container.libcint_bas[ATOM_OF  + BAS_SLOTS * n] = atom_index;
```
Comment (tmhuysen) [inline src/Basis/LibcintInterfacer.cpp]: first use `ATOM_OF` 
use of `BAS_SLOTS` after `bas_slots` declaration

File: src/Basis/LibcintInterfacer.cpp [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+
+        // Set shell-related data into the libcint 'basis' and into the libcint environment
+        raw_container.libcint_bas[ANG_OF    + BAS_SLOTS * n] = static_cast<int>(current_shell.get_l());  // angular momentum
```
Comment (tmhuysen) [inline src/Basis/LibcintInterfacer.cpp]: first use `ANG_OF`

File: src/Basis/LibcintInterfacer.cpp [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        raw_container.libcint_bas[NPRIM_OF  + BAS_SLOTS * n] = static_cast<int>(current_shell.contractionSize());  // number of primitives
```
Comment (tmhuysen) [inline src/Basis/LibcintInterfacer.cpp]: first use `NPRIM_OF`

File: src/Basis/LibcintInterfacer.cpp [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+        raw_container.libcint_bas[NCTR_OF   + BAS_SLOTS * n] = 1;  // apparently, the number of contractions is always 1 (I'm still not sure what the libcint number of contractions means)
```
Comment (tmhuysen) [inline src/Basis/LibcintInterfacer.cpp]: first use `NCTR_OF`

File: src/Basis/LibcintInterfacer.cpp [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        raw_container.libcint_bas[PTR_EXP   + BAS_SLOTS * n] = offset;  // pointer to the exponents of the shell inside the libcint environment
```
Comment (tmhuysen) [inline src/Basis/LibcintInterfacer.cpp]: `PTR_EXP`

File: src/Basis/LibcintInterfacer.cpp [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    const auto& atoms = shell_set.atoms();
+    const auto& natm = atoms.size();
+    const auto& nbf = shell_set.numberOfBasisFunctions();
+    const auto& nsh = shell_set.size();  // number of shells
+
+    libcint::RawContainer raw_container (natm, nbf, nsh);
+
+
+    // Configuration of atom-related data
+    int offset = PTR_ENV_START;  // an offset such that libcint can retrieve the correct index inside the environment, starts at 20
```
Comment (tmhuysen) [inline src/Basis/LibcintInterfacer.cpp]: `PTR_ENV_START` first use

File: src/Basis/LibcintInterfacer.cpp [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+
+    for (size_t i = 0; i < natm; i++) {
+        // Configure a libcint 'atom'
+        raw_container.libcint_atm[CHARGE_OF + ATM_SLOTS * i] = static_cast<int>(atoms[i].atomic_number);  // insert the charge/atomic number
```
Comment (tmhuysen) [inline src/Basis/LibcintInterfacer.cpp]: `CHARGE_OF`first use  , `ATM_SLOTS` use after declaration `atm_slots`

File: src/Basis/LibcintInterfacer.cpp [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+        raw_container.libcint_atm[PTR_COORD + ATM_SLOTS * i] = offset;  // pointer to the coordinates of the atom inside the libcint environment
```
Comment (tmhuysen) [inline src/Basis/LibcintInterfacer.cpp]: `PTR_COORD` first use, `ATM_SLOTS ` `atm_slots`

File: src/Basis/LibcintInterfacer.cpp (hunk 4 above)
Comment (tmhuysen) [inline src/Basis/LibcintInterfacer.cpp]: `bas_slots`

File: src/Basis/LibcintInterfacer.cpp (hunk 5 above)
Comment (tmhuysen) [inline src/Basis/LibcintInterfacer.cpp]: `bas_slots`

File: src/Basis/LibcintInterfacer.cpp (hunk 6 above)
Comment (tmhuysen) [inline src/Basis/LibcintInterfacer.cpp]: `bas_slots`

File: src/Basis/LibcintInterfacer.cpp (hunk 7 above)
Comment (tmhuysen) [inline src/Basis/LibcintInterfacer.cpp]: `bas_slots`

File: src/Basis/LibcintInterfacer.cpp [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+        raw_container.libcint_bas[NCTR_OF   + BAS_SLOTS * n] = 1;  // apparently, the number of contractions is always 1 (I'm still not sure what the libcint number of contractions means)
+        raw_container.libcint_bas[PTR_EXP   + BAS_SLOTS * n] = offset;  // pointer to the exponents of the shell inside the libcint environment
+
+        const auto& gaussian_exponents = current_shell.get_gaussian_exponents();
+        for (size_t e = 0; e < gaussian_exponents.size(); e++, offset++) {  // also increment offset
+            raw_container.libcint_env[offset] = gaussian_exponents[e];
+        }
+
+
+        raw_container.libcint_bas[PTR_COEFF + BAS_SLOTS * n] = offset;  // pointer to the contraction coefficients inside the libcint environment
```
Comment (tmhuysen) [inline src/Basis/LibcintInterfacer.cpp]: `PTR_COEFF`first use, `bas_slots`

File: src/Basis/LibcintInterfacer.cpp [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+        // Input NORMALIZED contraction coefficients inside the libcint 'env'
+        if (current_shell.are_embedded_normalization_factors_of_primitives()) {
+            current_shell.unEmbedNormalizationFactorsOfPrimitives();
+        }
+
+        const auto& current_contraction_coefficients = current_shell.get_contraction_coefficients();
+        for (size_t c = 0; c < current_contraction_coefficients.size(); c++, offset++) {  // also increment offset
+            raw_container.libcint_env[offset] = current_contraction_coefficients[c] * CINTgto_norm(raw_container.libcint_bas[ANG_OF+BAS_SLOTS*n], raw_container.libcint_env[raw_container.libcint_bas[PTR_EXP+BAS_SLOTS*n]+c]);  // use libcint to embed the norm of the primitives into the contraction coefficient
```
Comment (tmhuysen) [inline src/Basis/LibcintInterfacer.cpp]: `bas_slots`

File: include/Basis/LibcintInterfacer.hpp [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+
+
+
+extern "C" {
+
+#include <cint.h>
+
+
+/*
+ *  FUNCTIONS THAT AREN'T INSIDE <cint.h>
```
[the remark this answers is not in the document]
Comment (lelemmen) [reply include/Basis/LibcintInterfacer.hpp]: `<cint.h>` does not contain these functions, only the source files/library contains them. If we want to use the functions, we have to let the compiler know that they exist.

**Pull request merged**